CORE-526 k/configs: dont override topic-level cleanup policy

pgellert opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This fixes a bug in the kafka layer where redpanda would set the `cleanup.policy` config for every new topic even if this was not explicitly specified by the topic creation request. Instead, the expected behaviour is that this topic-level config is not set, but instead, redpanda should fall back in the case of an unset `cleanup.policy` to the cluster-level default without hardcoding that default at the topic-level.

Fixes https://redpandadata.atlassian.net/browse/CORE-526
Fixes https://redpandadata.atlassian.net/browse/CORE-2807


## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.1.x
- [ ] v23.3.x
- [ ] v23.2.x

## Release Notes

### Bug Fixes

* This fixes a bug in kafka topic configs, where the `cleanup.policy` config was always set at the topic-level to the cluster default even when the topic creation request did not specify this.

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

pgellert commented:
/ci-repeat 1

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/48786#018f53af-adaa-43bf-9249-2318001540b4:
```
"rptest.tests.topic_recovery_test.TopicRecoveryTest.test_size_based_retention.cloud_storage_type=CloudStorageType.ABS"
```
<!--5159f44a-dee2-4e57-ba9f-4b56dadcb206-->

pgellert commented:
/ci-repeat 1

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/48846#018f5e08-f600-4b46-9df2-0f80c8e34da5
<!--b10386f3-7285-4a99-9b1a-e6b9e6ab9c01-->

pgellert commented:
/ci-repeat 1

pgellert commented:
/ci-repeat 1

pgellert commented:
/ci-repeat 1

Lazin commented on src/v/cluster/partition_recovery_manager.cc:
```diff
@@ -245,13 +245,11 @@ std::ostream& operator<<(std::ostream& o, const retention& r) {
-static retention
-get_retention_policy(const storage::ntp_config::default_overrides& prop) {
-    auto flags = prop.cleanup_policy_bitflags;
+static retention get_retention_policy(const storage::ntp_config& prop) {
```
why this change is needed?

Lazin reviewed (commented):
You also need to update the place where we generate topic_manifest in the ntp_config. 
Also, when we update the cleanup policy flag in the topic config we have to restart archiver (in the `partition::update_configuration` the flag is set in this case). But with this change we also need to do the same when the configuration parameter changes.

pgellert commented:
[Force-push](https://github.com/redpanda-data/redpanda/compare/6c7d2ca4b8e8ad0f361cdb52d45c0d9782445700..c76b6c6d3ea3f036f0fd071a1b0bc7d99147e142): rebased to dev and addressed the feedback. (It's best to re-review the PR as a whole rather than looking at the force-push diff.)

pgellert commented:
@Lazin
> Also, when we update the cleanup policy flag in the topic config we have to restart archiver (in the `partition::update_configuration` the flag is set in this case). But with this change we also need to do the same when the configuration parameter changes.

I've done this now.

> You also need to update the place where we generate topic_manifest in the ntp_config.

As far as I can see the (v2) topic_manifest serializes the `cluster::topic_configuration` into the `topic_manifest`, ie. it only serializes the topic-specific overrides into the manifest rather than the "resolved value" (= topic specific override or else the cluster config based callback). 

By "update the place where we generate topic_manifest in the ntp_config", did you mean that the `topic_manifest` should contain the resolved value of the `cleanup_policy` config instead?

pgellert commented on src/v/cluster/partition.cc:
```diff
@@ -82,6 +83,23 @@ partition::partition(
         }
     }
+
+    _log_cleanup_policy.watch([this]() {
+        if (_as.abort_requested()) {
+            return ss::now();
+        }
+        auto changed = _raft->log()->notify_compaction_update();
+        if (changed) {
+            vlog(
+              clusterlog.debug,
+              "[{}] updating archiver for cluster config change in "
+              "log_cleanup_policy",
+              _raft->ntp());
+
+            return restart_archiver(false);
+        }
+        return ss::now();
+    });
```
@Lazin We're doing the archiver restart synchronously on topic config updates, so I also went with making the restart on config updates synchronous here. But I wonder if that's safe or if I should offload it to a fibre? What do you think?

pgellert commented:
/dt

pgellert commented:
[Force-pushed](https://github.com/redpanda-data/redpanda/compare/c76b6c6d3ea3f036f0fd071a1b0bc7d99147e142..c7faff39692f449a89dea10585b0f9b058a95eb4) to fix test compilation errors.

pgellert commented:
/dt

vbotbuildovich commented:
<!--NO CACHE-->
skipped ducktape retry in https://buildkite.com/redpanda/redpanda/builds/51145#01908324-653b-4590-8818-137cef15b65e:
pandatriage cache was not found
<!--602a9ecd-873b-4b7d-8d45-19942f8a1852-->

Lazin commented on src/v/cluster/partition.cc:
```diff
@@ -742,7 +742,7 @@ ss::future<> partition::update_configuration(topic_properties properties) {
       old_ntp_config.is_archival_enabled() != new_archival
       || old_ntp_config.is_read_replica_mode_enabled()
-           != new_ntp_config.read_replica
+           != new_ntp_config.read_replica.value_or(false)
```
it should be impossible to enable or disable read-replica mode, the normal topic can't be converted to read-replica and vice versa

pgellert replied:
Should I remove this check then instead of fixing it up? As I wrote in the commit message (https://github.com/redpanda-data/redpanda/pull/18284/commits/aff708585e723b25da1725ce5d93bab11ad9f128), it was causing archival restarts on unrelated changes when the config was set to `read_replica = std::optional(false)`.

pgellert replied:
Fixed

pgellert commented on src/v/cluster/partition.cc:
```diff
@@ -730,20 +730,40 @@ ss::future<> partition::update_configuration(topic_properties properties) {
+          "read_replica",
+          _raft->ntp());
+        cloud_storage_changed = true;
+    } else if (old_compaction_status != new_compaction_status) {
```
It could be just an `if`. `else if` ensures that we only log once for restarts and we log the first reason that caused a restart. It's possible to have multiple config changes in a single topic update that cause a restart, so I could make it an `if` to provide more context.

pgellert replied:
Fixed

pgellert commented:
[Force-pushed](https://github.com/redpanda-data/redpanda/compare/c7faff39692f449a89dea10585b0f9b058a95eb4..e18639988d597b1d43f8408c326296ac98c31273) to change the checks on topic property update - there's no need to check the read replica config for changes as it is immutable after topic creation. I also added back checks for changes to `retention.bytes` and `retention.ms`
as these configs are used by archiver and `CloudArchiveRetentionTest.test_delete` fails without restarting archiver whenever these configs change.

pgellert commented on src/v/cluster/partition.cc:
```diff
@@ -730,20 +730,40 @@ ss::future<> partition::update_configuration(topic_properties properties) {
                           new_ntp_config.shadow_indexing_mode.value());
 
+    bool old_read_replica = old_ntp_config.is_read_replica_mode_enabled();
+    bool new_read_replica = new_ntp_config.read_replica.value_or(false);
```
Fixed

BenPope commented on src/v/storage/disk_log_impl.h:
```diff
@@ -380,6 +381,7 @@ class disk_log_impl final : public log {
     std::optional<model::offset> _cloud_gc_offset;
     std::optional<model::offset> _last_compaction_window_start_offset;
     size_t _reclaimable_size_bytes{0};
+    bool _compaction_enabled;
```
Does `_compaction_enabled` need to exist? Does it ever diverge from `config().is_compacted()`?

pgellert replied:
Yes, this stores the current state of the log's compaction status and it's used to compare against `config().is_compacted()` for changes in `disk_log_impl::notify_compaction_update`. We can't just use `config().is_compacted()` because when the `log_cleanup_policy` cluster config changes and we try to determine whether we need to (un)mark the segments as compacted, we need a way to remember what the old compaction status was.

BenPope commented on src/v/storage/disk_log_impl.h:
```diff
@@ -146,7 +146,8 @@ class disk_log_impl final : public log {
     size_t size_bytes() const override { return _probe->partition_size(); }
     uint64_t size_bytes_after_offset(model::offset o) const override;
-    ss::future<> update_configuration(ntp_config::default_overrides) final;
+    void set_overrides(ntp_config::default_overrides) final;
+    bool notify_compaction_update() final;
```
I'm not sure why this was split, it seems easier to misuse now.

pgellert replied:
These two manage independent behaviours. When the cluster-level config `log_cleanup_policy` changes there is no `ntp_config::default_overrides` to update but the compaction status still needs to update. I was following the pattern used in the consensus layer (sitting between `partition` and `log`) which has a similar `notify_config_update` method for write caching config changes:
https://github.com/redpanda-data/redpanda/blob/1e1557564a4990de37434e36cb1e6baaeaf5f1d2/src/v/raft/consensus.h#L516-L517
It's only used in `partition.cc` and partition+consensus+log are already highly coupled, so I think this is fine. (Perhaps the ideal way would be to have a methot with a signature `ss::future<change_result> apply_config_change(std::variant<topic_config_change, cluster_config_change>);` but I wanted to keep the refactoring minimal and consistent with the existing pattern.)

BenPope replied:
> (Perhaps the ideal way would be to have a methot with a signature ss::future<change_result> apply_config_change(std::variant<topic_config_change, cluster_config_change>); but I wanted to keep the refactoring minimal and consistent with the existing pattern.)

The concern here is that splitting it up means that the two calls must both be made or there's a bug. Within the log, it's not clear whether to read `_compaction_enabled` or `config().is_compacted()`.

It might be worth changing the name, or making it an optional that's only engaged during the gap between the two calls.

BenPope commented on src/v/cluster/partition.cc:
```diff
@@ -84,6 +85,23 @@ partition::partition(
+        if (_as.abort_requested()) {
+            return ss::now();
+        }
+        auto changed = _raft->log()->notify_compaction_update();
```
Can the log not manage this itself, with the config passed through log_manager?

pgellert replied:
I could have a watcher in `log_manager` that calls `log->notify_compaction_update()`, but (1) we would still need `partition` to know about changes to the compaction status to ensure that `archiver` is restarted and (2) it seems that `partition` already manages the compaction status of the log through `partition::update_configuration`.

pgellert commented on src/v/cluster/partition.cc:
```diff
@@ -752,11 +756,35 @@ ss::future<> partition::update_configuration(topic_properties properties) {
+        vlog(
+          clusterlog.debug,
+          "[{}] updating archiver for topic config change in compaction",
+          _raft->ntp());
```
That's what I did initially but then @Lazin recommended splitting it: https://github.com/redpanda-data/redpanda/pull/18284#discussion_r1668677290. I think I'm going to stay with the current approach for now.

BenPope replied:
> That's what I did initially but then @Lazin recommended splitting it: [#18284 (comment)](https://github.com/redpanda-data/redpanda/pull/18284#discussion_r1668677290). I think I'm going to stay with the current approach for now.

It looks like initially you had it printing one, not all? Anyway, I'm happy to take @Lazin's suggestion.

pgellert commented:
[Force-pushed](https://github.com/redpanda-data/redpanda/compare/e18639988d597b1d43f8408c326296ac98c31273..800d4e7a8e9412b85fff6fa77075be668673df69) to address some nitpicks (using the new helpers in more places).

pgellert commented on src/v/model/fundamental.h:
```diff
@@ -129,6 +129,16 @@ inline void operator&=(cleanup_policy_bitflags& a, cleanup_policy_bitflags b) {
 }
 
+inline bool is_compaction_enabled(cleanup_policy_bitflags flags) {
+    return (flags & cleanup_policy_bitflags::compaction)
```
Done

dotnwat commented:
Should this behavior cahnge be backported, or should be enabled only in new clusters with an appropriate announcement in the release notes?

pgellert commented:
Good point. I am going to hold off on backporting it for now and talk to devex/product to see if there is a strong motivation to backport it.

That being said, I consider this a bug fix. Existing topics will have their `cleanup.policy` set as an override and it is only when they create a new topic or issue an AlterConfigs request that a topic's cleanup policy will change.

I put an announcement in the release notes up in the PR cover letter. Or did you have anything more in mind?

pgellert commented:
@dotnwat I had a chat with @michael-redpanda and we think of this change as a bug fix that we should backport. Backporting this resolves a recent bug report: https://github.com/redpanda-data/redpanda/issues/21360

pgellert commented:
/backport v23.3.x

vbotbuildovich commented:
Failed to create a backport PR to v24.1.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-18284-v24.1.x-970 remotes/upstream/v24.1.x
git cherry-pick -x 9a3f84695eaee59f096b7eeb8fbba8caf4d82523 6a26dba9b73f3f59fbe5d25a42bf62a3446426a1 f2bea7fbc7488670733afe54ec2f71a93632f4a3 c3e2be3042a362ca5a12eeed63242732914619c5 b4ac3d76c37ec1d5eb4ee93374881ac09068503f 35fb002268d75a51490b70d003072c72d64a6741 7517614b708e7d645ea0993c60b51b1160a161a0 811a5edd60d4c3c30f1efa343d6e07e58f4f68d8 2504b4c32c6c9cd70e7ee8a1a1149f78d7ea8eb2 f24256a98c0142ae460fd1e18a94a78e39ffe6b0 800d4e7a8e9412b85fff6fa77075be668673df69
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/9943335009).

vbotbuildovich commented:
Failed to create a backport PR to v23.3.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-18284-v23.3.x-521 remotes/upstream/v23.3.x
git cherry-pick -x 9a3f84695eaee59f096b7eeb8fbba8caf4d82523 6a26dba9b73f3f59fbe5d25a42bf62a3446426a1 f2bea7fbc7488670733afe54ec2f71a93632f4a3 c3e2be3042a362ca5a12eeed63242732914619c5 b4ac3d76c37ec1d5eb4ee93374881ac09068503f 35fb002268d75a51490b70d003072c72d64a6741 7517614b708e7d645ea0993c60b51b1160a161a0 811a5edd60d4c3c30f1efa343d6e07e58f4f68d8 2504b4c32c6c9cd70e7ee8a1a1149f78d7ea8eb2 f24256a98c0142ae460fd1e18a94a78e39ffe6b0 800d4e7a8e9412b85fff6fa77075be668673df69
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/9943339893).
